[IsDeprecatedWeakRefSmartPointerException] Make HTTPCookieStoreObserver RefCounted

basuke opened the pull request:
#### a6b0a2d42b02a548c67c94065f64799faa394cf3
<pre>
[IsDeprecatedWeakRefSmartPointerException] Make HTTPCookieStoreObserver RefCounted
<a href="https://bugs.webkit.org/show_bug.cgi?id=280072">https://bugs.webkit.org/show_bug.cgi?id=280072</a>

Reviewed by Geoffrey Garen.

Every CanMakeWeakPtr class should also RefCounted.

* Source/WebKit/UIProcess/API/APIHTTPCookieStore.h:
* Source/WebKit/UIProcess/API/APIHTTPCookieStore.mm:
(HTTPCookieStore::cookiesDidChange)
* Source/WebKit/UIProcess/API/Cocoa/WKHTTPCookieStore.mm:
(WKHTTPCookieStoreObserver::create):
(-[WKHTTPCookieStore dealloc]):
(-[WKHTTPCookieStore addObserver:]):
* Source/WebKit/UIProcess/API/glib/WebKitCookieManager.cpp:
(CookieStoreObserver::create):
(webkitCookieManagerCreate):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm:
(WebKit::WebExtensionController::addWebsiteDataStore):
* Source/WebKit/UIProcess/Extensions/WebExtensionController.h:
(WebKit::WebExtensionController::HTTPCookieStoreObserver::create):

Canonical link: <a href="https://commits.webkit.org/284008@main">https://commits.webkit.org/284008@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fcb5b9775ea2cdf61d86dd7db26c8a7fe0c2eb33

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/68062 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/47454 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/20721 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/72125 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/19207 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/55250 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/19023 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/54401 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/12809 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/71129 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/43454 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/58813 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/34863 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/40123 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/17564 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/62094 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/16570 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/73821 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/12033 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/15854 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/61850 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/12071 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/58888 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/61866 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15123 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/9771 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/3383 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/43255 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/44328 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/45523 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/44070 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/UIProcess/API/APIHTTPCookieStore.h:
```diff
@@ -60,7 +52,7 @@ namespace API {
 class HTTPCookieStore;
 
-class HTTPCookieStoreObserver : public CanMakeWeakPtr<HTTPCookieStoreObserver> {
+class HTTPCookieStoreObserver : public RefCountedAndCanMakeWeakPtr<HTTPCookieStoreObserver> {
```
It is unfortunate to go through the pain to make this RefCounted but then not leverage it at all to improve safety. This PR should include the following improvement:
```
void HTTPCookieStore::cookiesDidChange()
{
    for (auto& observer : m_observers)
        observer.cookiesDidChange(*this);
}
```
->
```
void HTTPCookieStore::cookiesDidChange()
{
    for (Ref observer : m_observers)
        observer->cookiesDidChange(*this);
}
```
in APIHttpCookieStore.cpp.

basuke replied:
Oh, right. Thanks. I really want the static analyzer in EWS.

geoffreygaren reviewed (approved):
r=me

basuke committed:
[IsDeprecatedWeakRefSmartPointerException] Make HTTPCookieStoreObserver RefCounted
https://bugs.webkit.org/show_bug.cgi?id=280072

Reviewed by Geoffrey Garen.

Every CanMakeWeakPtr class should also RefCounted.

* Source/WebKit/UIProcess/API/APIHTTPCookieStore.h:
* Source/WebKit/UIProcess/API/APIHTTPCookieStore.mm:
(HTTPCookieStore::cookiesDidChange)
* Source/WebKit/UIProcess/API/Cocoa/WKHTTPCookieStore.mm:
(WKHTTPCookieStoreObserver::create):
(-[WKHTTPCookieStore dealloc]):
(-[WKHTTPCookieStore addObserver:]):
* Source/WebKit/UIProcess/API/glib/WebKitCookieManager.cpp:
(CookieStoreObserver::create):
(webkitCookieManagerCreate):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm:
(WebKit::WebExtensionController::addWebsiteDataStore):
* Source/WebKit/UIProcess/Extensions/WebExtensionController.h:
(WebKit::WebExtensionController::HTTPCookieStoreObserver::create):

Canonical link: https://commits.webkit.org/284008@main
